pallets/oracle-data-collection/src/lib.rs

```diff
@@ -0,0 +1,496 @@
+
+	pub type OracleValuePair<T> = (<T as Config>::OracleValue, <T as Config>::Timestamp);
+
+	/// Type containing the associated info to a key
+	#[derive(Encode, Decode, Clone, PartialEq, Eq, TypeInfo, RuntimeDebug, MaxEncodedLen)]
+	#[scale_info(skip_type_params(T))]
+	pub struct KeyInfo<T: Config> {
+		pub feeders: BoundedVec<T::FeederId, T::MaxFeedersPerKey>,
+		pub usage_refs: u32,
+	}
+
+	impl<T: Config> Default for KeyInfo<T> {
+		fn default() -> Self {
+			Self {
+				feeders: Default::default(),
+				usage_refs: 0,
+			}
+		}
+	}
+
+	impl<T: Config> KeyInfo<T> {
+		pub fn is_clean(&self) -> bool {
+			self.feeders.is_empty() && self.usage_refs.is_zero()
+		}
+	}
+
+	/// A collection cached in memory
+	#[derive(Clone)]
+	pub struct CachedCollection<T: Config>(
+		pub BoundedBTreeMap<T::OracleKey, OracleValuePair<T>, T::MaxCollectionSize>,
+	);
+
+	impl<T: Config> DataCollection<T::OracleKey> for CachedCollection<T> {
+		type Data = OracleValuePair<T>;
+
+		fn get(&self, data_id: &T::OracleKey) -> Result<OracleValuePair<T>, DispatchError> {
+			self.0
+				.get(data_id)
+				.cloned()
+				.ok_or_else(|| Error::<T>::KeyNotInCollection.into())
+		}
+	}
+
+	impl<T: Config> CachedCollection<T> {
+		pub fn as_vec(self) -> Vec<(T::OracleKey, OracleValuePair<T>)> {
+			self.0.into_iter().collect()
+		}
+	}
+
+	/// Change done through a change guard.
+	#[derive(Encode, Decode, Clone, PartialEq, Eq, TypeInfo, RuntimeDebug, MaxEncodedLen)]
+	#[scale_info(skip_type_params(T))]
+	pub enum Change<T: Config> {
+		Feeders(T::OracleKey, BoundedVec<T::FeederId, T::MaxFeedersPerKey>),
+	}
+}
+
+/// Traits specifically used by this pallet
+pub mod traits {
+	/// Defined an aggregation behavior
+	pub trait AggregationProvider<Value, Timestamp> {
+		fn aggregate(
+			pairs: impl IntoIterator<Item = (Value, Timestamp)>,
+		) -> Option<(Value, Timestamp)>;
+	}
+}
+
+/// Provide types to use in runtime to configure this pallet
+pub mod util {
+	use sp_std::vec::Vec;
+
+	use super::traits::AggregationProvider;
+
+	/// Type that performs an aggregation using the median for values and
+	/// timestamps
+	pub struct MedianAggregation;
+
+	impl<Value, Timestamp> AggregationProvider<Value, Timestamp> for MedianAggregation
+	where
+		Value: Ord + Clone,
+		Timestamp: Ord + Clone,
+	{
+		fn aggregate(
+			pairs: impl IntoIterator<Item = (Value, Timestamp)>,
+		) -> Option<(Value, Timestamp)> {
+			let (mut values, mut timestamps): (Vec<_>, Vec<_>) = pairs.into_iter().unzip();
+
+			let value = median(&mut values)?.clone();
+			let timestamp = median(&mut timestamps)?.clone();
+
+			Some((value, timestamp))
+		}
+	}
+
+	/// Computes fastly the median of a list of values
+	/// Extracted from orml
+	pub fn median<T: Ord>(items: &mut Vec<T>) -> Option<&T> {
+		if items.is_empty() {
+			return None;
+		}
+
+		let mid_index = items.len() / 2;
+
+		// Won't panic as `items` ensured not empty.
+		let (_, item, _) = items.select_nth_unstable(mid_index);
+		Some(item)
+	}
```

Comment:
Q: Why can't we just use `items.get(items.len() / 2)` which returns `None` by default if the index is out of bounds?